File: internal/protocol/value.go
package protocol

import (
	"bytes"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"math"
	"sort"
	"strconv"
	"strings"
	"time"
	"unicode/utf8"

	"github.com/google/cel-go/common/types"
	"github.com/google/cel-go/common/types/ref"
	"github.com/google/cel-go/common/types/traits"
)

const maxValueDepth = 32
const maxCollectionItems = 4096

func DecodeVariables(raw string, maxBytes, maxDepth int) (map[string]any, error) {
	if len(raw) > maxBytes {
		return nil, fmt.Errorf("variables JSON exceeds %d bytes", maxBytes)
	}
	if !utf8.ValidString(raw) {
		return nil, fmt.Errorf("variables JSON must be valid UTF-8")
	}
	decoder := json.NewDecoder(strings.NewReader(raw))
	decoder.UseNumber()
	var decoded any
	if err := decoder.Decode(&decoded); err != nil {
		return nil, fmt.Errorf("invalid variables JSON: %w", err)
	}
	var trailing any
	if err := decoder.Decode(&trailing); err != io.EOF {
		if err == nil {
			return nil, fmt.Errorf("variables JSON contains trailing data")
		}
		return nil, fmt.Errorf("invalid trailing variables JSON: %w", err)
	}
	object, ok := decoded.(map[string]any)
	if !ok {
		return nil, fmt.Errorf("variables JSON must be an object")
	}
	value, err := decodeInput(object, 0, maxDepth)
	if err != nil {
		return nil, err
	}
	return value.(map[string]any), nil
}

func decodeInput(value any, depth, maxDepth int) (any, error) {
	if depth > maxDepth {
		return nil, fmt.Errorf("variables nesting exceeds %d levels", maxDepth)
	}
	switch value := value.(type) {
	case nil, bool, string:
		return value, nil
	case json.Number:
		return decodeNumber(value)
	case []any:
		if len(value) > maxCollectionItems {
			return nil, fmt.Errorf("list exceeds %d items", maxCollectionItems)
		}
		result := make([]any, len(value))
		for i, item := range value {
			decoded, err := decodeInput(item, depth+1, maxDepth)
			if err != nil {
				return nil, err
			}
			result[i] = decoded
		}
		return result, nil
	case map[string]any:
		if tagged, ok := decodeTagged(value); ok {
			return tagged, nil
		}
		if len(value) > maxCollectionItems {
			return nil, fmt.Errorf("map exceeds %d entries", maxCollectionItems)
		}
		result := make(map[string]any, len(value))
		for key, item := range value {
			decoded, err := decodeInput(item, depth+1, maxDepth)
			if err != nil {
				return nil, err
			}
			result[key] = decoded
		}
		return result, nil
	default:
		return nil, fmt.Errorf("unsupported JSON value %T", value)
	}
}

func decodeNumber(value json.Number) (any, error) {
	text := value.String()
	if strings.ContainsAny(text, ".eE") {
		decoded, err := strconv.ParseFloat(text, 64)
		if err != nil || math.IsInf(decoded, 0) || math.IsNaN(decoded) {
			return nil, fmt.Errorf("invalid JSON number %q", text)
		}
		return decoded, nil
	}
	if decoded, err := strconv.ParseInt(text, 10, 64); err == nil {
		return decoded, nil
	}
	if decoded, err := strconv.ParseUint(text, 10, 64); err == nil {
		return decoded, nil
	}
	return nil, fmt.Errorf("integer %q is outside CEL int/uint range", text)
}

func decodeTagged(value map[string]any) (any, bool) {
	kind, ok := value["kind"].(string)
	if !ok {
		return nil, false
	}
	raw := value["value"]
	switch kind {
	case "null":
		return nil, true
	case "bool":
		decoded, ok := raw.(bool)
		return decoded, ok
	case "int":
		text, ok := raw.(string)
		if !ok {
			return nil, false
		}
		decoded, err := strconv.ParseInt(text, 10, 64)
		return decoded, err == nil
	case "uint":
		text, ok := raw.(string)
		if !ok {
			return nil, false
		}
		decoded, err := strconv.ParseUint(text, 10, 64)
		return decoded, err == nil
	case "double":
		text, ok := raw.(string)
		if !ok {
			return nil, false
		}
		decoded, err := parseDouble(text)
		return decoded, err == nil
	case "string":
		decoded, ok := raw.(string)
		return decoded, ok
	case "bytes":
		text, ok := raw.(string)
		if !ok {
			return nil, false
		}
		decoded, err := base64.StdEncoding.DecodeString(text)
		return decoded, err == nil
	case "timestamp":
		text, ok := raw.(string)
		if !ok {
			return nil, false
		}
		decoded, err := time.Parse(time.RFC3339Nano, text)
		return decoded.UTC(), err == nil
	case "duration":
		text, ok := raw.(string)
		if !ok {
			return nil, false
		}
		decoded, err := time.ParseDuration(text)
		return decoded, err == nil
	default:
		return nil, false
	}
}

func parseDouble(value string) (float64, error) {
	switch value {
	case "NaN":
		return math.NaN(), nil
	case "Infinity", "+Infinity", "+Inf":
		return math.Inf(1), nil
	case "-Infinity", "-Inf":
		return math.Inf(-1), nil
	default:
		return strconv.ParseFloat(value, 64)
	}
}

func EncodeValue(value ref.Val) (Value, error) {
	return encodeValue(value, 0)
}

func encodeValue(value ref.Val, depth int) (Value, error) {
	if value == nil {
		return Value{}, fmt.Errorf("nil CEL value")
	}
	if depth > maxValueDepth {
		return Value{}, fmt.Errorf("CEL value nesting exceeds %d levels", maxValueDepth)
	}
	switch value := value.(type) {
	case types.Null:
		return Value{Kind: "null"}, nil
	case types.Bool:
		return Value{Kind: "bool", Value: bool(value)}, nil
	case types.Int:
		return Value{Kind: "int", Value: strconv.FormatInt(int64(value), 10)}, nil
	case types.Uint:
		return Value{Kind: "uint", Value: strconv.FormatUint(uint64(value), 10)}, nil
	case types.Double:
		return Value{Kind: "double", Value: formatDouble(float64(value))}, nil
	case types.String:
		return Value{Kind: "string", Value: string(value)}, nil
	case types.Bytes:
		return Value{Kind: "bytes", Value: base64.StdEncoding.EncodeToString([]byte(value))}, nil
	case types.Timestamp:
		return Value{Kind: "timestamp", Value: value.Time.UTC().Format(time.RFC3339Nano)}, nil
	case types.Duration:
		return Value{Kind: "duration", Value: formatDuration(value.Duration)}, nil
	}
	if list, ok := value.(traits.Lister); ok {
		return encodeList(list, depth)
	}
	if mapper, ok := value.(traits.Mapper); ok {
		return encodeMap(mapper, depth)
	}
	return Value{}, fmt.Errorf("unsupported CEL value type %q", value.Type().TypeName())
}

func encodeList(list traits.Lister, depth int) (Value, error) {
	size, ok := list.Size().(types.Int)
	if !ok || size < 0 || size > maxCollectionItems {
		return Value{}, fmt.Errorf("unsupported CEL list size")
	}
	items := make([]Value, int(size))
	for i := range items {
		item, err := encodeValue(list.Get(types.Int(i)), depth+1)
		if err != nil {
			return Value{}, err
		}
		items[i] = item
	}
	return Value{Kind: "list", Items: items}, nil
}

func encodeMap(mapper traits.Mapper, depth int) (Value, error) {
	size, ok := mapper.Size().(types.Int)
	if !ok || size < 0 || size > maxCollectionItems {
		return Value{}, fmt.Errorf("unsupported CEL map size")
	}
	entries := make([]MapEntry, 0, int(size))
	iterator := mapper.Iterator()
	for iterator.HasNext() == types.True {
		key := iterator.Next()
		value, found := mapper.Find(key)
		if !found {
			return Value{}, fmt.Errorf("map value disappeared during iteration")
		}
		encodedKey, err := encodeValue(key, depth+1)
		if err != nil {
			return Value{}, err
		}
		encodedValue, err := encodeValue(value, depth+1)
		if err != nil {
			return Value{}, err
		}
		entries = append(entries, MapEntry{Key: encodedKey, Value: encodedValue})
	}
	sort.Slice(entries, func(i, j int) bool {
		left, _ := json.Marshal(entries[i].Key)
		right, _ := json.Marshal(entries[j].Key)
		return bytes.Compare(left, right) < 0
	})
	return Value{Kind: "map", Entries: entries}, nil
}

func formatDouble(value float64) string {
	switch {
	case math.IsNaN(value):
		return "NaN"
	case math.IsInf(value, 1):
		return "Infinity"
	case math.IsInf(value, -1):
		return "-Infinity"
	default:
		return strconv.FormatFloat(value, 'g', -1, 64)
	}
}

func formatDuration(value time.Duration) string {
	if value == 0 {
		return "0s"
	}
	sign := ""
	magnitude := uint64(value)
	if value < 0 {
		sign = "-"
		magnitude = uint64(-(value + 1)) + 1
	}
	seconds := magnitude / uint64(time.Second)
	nanos := magnitude % uint64(time.Second)
	return fmt.Sprintf("%s%d.%09ds", sign, seconds, nanos)
}

File: sdk/rust/src/value.rs
use base64::{Engine as _, engine::general_purpose::STANDARD};
use serde_json::{Map, Value};

#[derive(Clone, Debug)]
pub enum CelValue {
    Null,
    Bool(bool),
    Int(i64),
    Uint(u64),
    Double(f64),
    String(String),
    Bytes(Vec<u8>),
    Timestamp(String),
    Duration(String),
    List(Vec<CelValue>),
    Map(Vec<CelMapEntry>),
}

#[derive(Clone, Debug, PartialEq)]
pub struct CelMapEntry {
    pub key: CelValue,
    pub value: CelValue,
}

impl PartialEq for CelValue {
    fn eq(&self, other: &Self) -> bool {
        match (self, other) {
            (Self::Null, Self::Null) => true,
            (Self::Bool(left), Self::Bool(right)) => left == right,
            (Self::Int(left), Self::Int(right)) => left == right,
            (Self::Uint(left), Self::Uint(right)) => left == right,
            (Self::Double(left), Self::Double(right)) => {
                (left.is_nan() && right.is_nan()) || left == right
            }
            (Self::String(left), Self::String(right)) => left == right,
            (Self::Bytes(left), Self::Bytes(right)) => left == right,
            (Self::Timestamp(left), Self::Timestamp(right)) => left == right,
            (Self::Duration(left), Self::Duration(right)) => left == right,
            (Self::List(left), Self::List(right)) => left == right,
            (Self::Map(left), Self::Map(right)) => left == right,
            _ => false,
        }
    }
}

impl CelValue {
    pub fn from_json(value: &Value) -> Result<Self, String> {
        let object = value
            .as_object()
            .ok_or_else(|| "CEL value must be an object".to_string())?;
        let kind = object
            .get("kind")
            .and_then(Value::as_str)
            .ok_or_else(|| "CEL value.kind must be a string".to_string())?;
        match kind {
            "null" => Ok(Self::Null),
            "bool" => Ok(Self::Bool(
                required(object, "value")?
                    .as_bool()
                    .ok_or_else(|| "bool.value must be a boolean".to_string())?,
            )),
            "int" => Ok(Self::Int(
                required_string(object, "value")?
                    .parse()
                    .map_err(|_| "int.value is outside the signed 64-bit range".to_string())?,
            )),
            "uint" => Ok(Self::Uint(
                required_string(object, "value")?
                    .parse()
                    .map_err(|_| "uint.value is outside the unsigned 64-bit range".to_string())?,
            )),
            "double" => Ok(Self::Double(parse_double(required_string(
                object, "value",
            )?)?)),
            "string" => Ok(Self::String(required_string(object, "value")?.to_string())),
            "bytes" => Ok(Self::Bytes(
                STANDARD
                    .decode(required_string(object, "value")?)
                    .map_err(|error| format!("invalid bytes.value: {error}"))?,
            )),
            "timestamp" => Ok(Self::Timestamp(
                required_string(object, "value")?.to_string(),
            )),
            "duration" => Ok(Self::Duration(
                required_string(object, "value")?.to_string(),
            )),
            "list" => Ok(Self::List(
                required(object, "items")?
                    .as_array()
                    .ok_or_else(|| "list.items must be an array".to_string())?
                    .iter()
                    .map(Self::from_json)
                    .collect::<Result<_, _>>()?,
            )),
            "map" => Ok(Self::Map(
                required(object, "entries")?
                    .as_array()
                    .ok_or_else(|| "map.entries must be an array".to_string())?
                    .iter()
                    .map(|entry| {
                        let entry = entry
                            .as_object()
                            .ok_or_else(|| "map entry must be an object".to_string())?;
                        Ok(CelMapEntry {
                            key: Self::from_json(required(entry, "key")?)?,
                            value: Self::from_json(required(entry, "value")?)?,
                        })
                    })
                    .collect::<Result<_, String>>()?,
            )),
            other => Err(format!("unsupported CEL value kind {other}")),
        }
    }

    pub fn to_json(&self) -> Value {
        let mut object = Map::new();
        object.insert("kind".to_string(), Value::String(self.kind().to_string()));
        match self {
            Self::Null => {}
            Self::Bool(value) => {
                object.insert("value".to_string(), Value::Bool(*value));
            }
            Self::Int(value) => {
                object.insert("value".to_string(), Value::String(value.to_string()));
            }
            Self::Uint(value) => {
                object.insert("value".to_string(), Value::String(value.to_string()));
            }
            Self::Double(value) => {
                object.insert("value".to_string(), Value::String(format_double(*value)));
            }
            Self::String(value) => {
                object.insert("value".to_string(), Value::String(value.clone()));
            }
            Self::Bytes(value) => {
                object.insert("value".to_string(), Value::String(STANDARD.encode(value)));
            }
            Self::Timestamp(value) | Self::Duration(value) => {
                object.insert("value".to_string(), Value::String(value.clone()));
            }
            Self::List(values) => {
                object.insert(
                    "items".to_string(),
                    Value::Array(values.iter().map(Self::to_json).collect()),
                );
            }
            Self::Map(entries) => {
                object.insert(
                    "entries".to_string(),
                    Value::Array(
                        entries
                            .iter()
                            .map(|entry| {
                                serde_json::json!({
                                    "key": entry.key.to_json(),
                                    "value": entry.value.to_json(),
                                })
                            })
                            .collect(),
                    ),
                );
            }
        }
        Value::Object(object)
    }

    fn kind(&self) -> &'static str {
        match self {
            Self::Null => "null",
            Self::Bool(_) => "bool",
            Self::Int(_) => "int",
            Self::Uint(_) => "uint",
            Self::Double(_) => "double",
            Self::String(_) => "string",
            Self::Bytes(_) => "bytes",
            Self::Timestamp(_) => "timestamp",
            Self::Duration(_) => "duration",
            Self::List(_) => "list",
            Self::Map(_) => "map",
        }
    }
}

fn required<'a>(object: &'a Map<String, Value>, name: &str) -> Result<&'a Value, String> {
    object
        .get(name)
        .ok_or_else(|| format!("CEL value is missing {name}"))
}

fn required_string<'a>(object: &'a Map<String, Value>, name: &str) -> Result<&'a str, String> {
    required(object, name)?
        .as_str()
        .ok_or_else(|| format!("{name} must be a string"))
}

fn parse_double(value: &str) -> Result<f64, String> {
    match value {
        "NaN" => Ok(f64::NAN),
        "Infinity" | "+Infinity" => Ok(f64::INFINITY),
        "-Infinity" => Ok(f64::NEG_INFINITY),
        other => other
            .parse()
            .map_err(|_| format!("invalid double value {other}")),
    }
}

fn format_double(value: f64) -> String {
    if value.is_nan() {
        "NaN".to_string()
    } else if value == f64::INFINITY {
        "Infinity".to_string()
    } else if value == f64::NEG_INFINITY {
        "-Infinity".to_string()
    } else {
        value.to_string()
    }
}
